src/detect-parse.c

```diff
@@ -1796,6 +1796,51 @@ static Signature *SigInitHelper(DetectEngineCtx *de_ctx, const char *sigstr,
     return NULL;
 }
 
+/**
+ * \brief Checks if a signature has the same source and destination
+ * \param s parsed signature
+ *
+ *  \retval 1 if source and destination are the sane, 0 otherwise
+ */
+static int SigHasSameSourceAndDestination(const Signature *s)
+{
+    int ret = 0;
+
+    if (!(s->flags & SIG_FLAG_SP_ANY) || !(s->flags & SIG_FLAG_DP_ANY)) {
+        if (DetectPortCmp(s->sp, s->dp) != 1) {
+            goto end;
+        }
+    }
+
+    if (!(s->flags & SIG_FLAG_SRC_ANY) || !(s->flags & SIG_FLAG_DST_ANY)) {
+        DetectAddress *src_v4 = s->init_data->src->ipv4_head;
+        DetectAddress *dst_v4 = s->init_data->dst->ipv4_head;
+        DetectAddress *src_v6 = s->init_data->src->ipv6_head;
+        DetectAddress *dst_v6 = s->init_data->dst->ipv6_head;
+
+        if (src_v4 != NULL && dst_v4 != NULL) {
+            if (DetectAddressCmp(src_v4, dst_v4) != ADDRESS_EQ) {
```

Comment:
both src and dst can be a list of DetectAddress objects, but DetectAddressCmp doesn't compare the list but just a single object.